go/adbc/driver/flightsql/flightsql_statement.go

```diff
@@ -485,13 +530,91 @@ func (s *statement) ExecutePartitions(ctx context.Context) (*arrow.Schema, adbc.
 
 	var (
 		info *flight.FlightInfo
+		poll *flight.PollInfo
 		out  adbc.Partitions
 		sc   *arrow.Schema
 		err  error
 	)
 
 	var header, trailer metadata.MD
-	if s.prepared != nil {
+	if s.incrementalState != nil {
+		if s.incrementalState.complete {
+			schema := s.incrementalState.schema
+			totalRecords := s.incrementalState.previousInfo.TotalRecords
+			// Reset the statement for reuse
+			s.incrementalState = &incrementalState{}
+			s.progress = 0.0
+			return schema, adbc.Partitions{}, totalRecords, nil
+		}
+
+		backoff := 100 * time.Millisecond
+		for {
+			// Keep polling until the query completes or we get new partitions
+			if s.prepared != nil {
+				poll, err = s.prepared.ExecutePoll(ctx, s.incrementalState.retryDescriptor, grpc.Header(&header), grpc.Trailer(&trailer), s.timeouts)
+			} else {
+				poll, err = s.query.poll(ctx, s.cnxn, s.incrementalState.retryDescriptor, grpc.Header(&header), grpc.Trailer(&trailer), s.timeouts)
+			}
+			if err != nil {
+				break
+			}
+			info = poll.GetInfo()
+			if info == nil {
+				// The server is misbehaving
+				// XXX: should we also issue a query cancellation?
+				s.incrementalState = &incrementalState{}
+				s.progress = 0.0
+				return nil, adbc.Partitions{}, -1, adbc.Error{
+					Msg:  "[Flight SQL] Server returned a PollInfo with no FlightInfo",
+					Code: adbc.StatusInternal,
+				}
+			}
+			info = &flight.FlightInfo{
```

Comment:
Could we just use https://pkg.go.dev/google.golang.org/protobuf@v1.32.0/proto#Clone instead?